### ISSUE TITLE
RETRACTED: 2-site CTM 'non-convergence' was an under-converged SU fixture (tau=2), not a CTM defect

### ISSUE THREAD

**Issue by yingjerkao** [tenax-lab/tenax#838]
> 🤖 **AI-generated issue** — written by Claude Code, posted by @yingjerkao.

On a physical D=2 Heisenberg simple-update state, the 2-site CTM fails to reach a fixed point at every χ except 4, and at two points returns an energy **below the exact ground state** — a number no normalized state can produce.

## The variational breach

`heisenberg_gate()` is exactly `S·S` for spin-½ (eigenvalues −0.75 singlet, +0.25 triplet ×3, J=1), and `compute_energy_ctm_2site` returns energy **per site** (`E_h + E_v`, one bond of each type per site). Both verified directly, not assumed. The exact square-lattice spin-½ AFM value is **−0.6694/site**, and `sublattice_rotate_gate` is an exact mapping, so it is the applicable bound.

Two points fall below it:

| χ | max_iter | E/site | vs −0.6694 |
|---|---|---|---|
| 16 | 1600 | **−1.006769264296370** | −0.337 (50% past the bound) |
| 24 | 400 | **−0.696674922299055** | −0.027 |

These are not merely inaccurate energies. A variational bound violation means the environment is not a valid reduced density matrix, so the quantity is not an expectation value of any state. Consequence: a result taken from this path can be unphysical rather than imprecise, and "run more sweeps" does not fix it — χ=16 was *inside* the bound at 400 sweeps and left it at 1600.

## Non-convergence

Fixed state, `conv_tol=1e-12`, only the iteration budget varies:

```
 chi  max_iter                      E   dE vs prev  rank(C1_A)  rank(C1_B)     wall
   4  100..3200  -0.271939907383136   +0.000e+00        4          4     converged
   6       400   -0.054896860564796                     6          6      0.2s
   6      1600   -0.574697207135369   -5.198e-01        6          6      0.5s
   6      3200   -0.532024925409080   +4.267e-02        6          6      1.0s
   6      6400   -0.537804976084891   -5.780e-03        6          6      1.7s
   8       100   -0.278252017903064                     8          8
   8       200   -0.544418759792349   -2.662e-01        8          8
   8      3200   -0.548978636050893   +1.723e-02        8          8
  12       400   -0.404973436382669                    12         12      0.9s
  12      1600   -0.538604374348900   -1.336e-01       12         12      1.5s
  12      3200   -0.547029990953815   -8.426e-03       12         12      2.8s
  12      6400   -0.496469663176715   +5.056e-02       12         12      5.5s
  16       400   -0.537139788204294                    15         16      1.1s
  16      1600   -1.006769264296370   -4.696e-01       16         16      2.3s
  16      3200   -0.548562320067812   +4.582e-01       16         16      4.7s
  16      6400   -0.545394746953308   +3.168e-03       16         16      8.9s
  24       400   -0.696674922299055                    24         24     89.8s
  24      1600   -0.537892204430865   +1.588e-01       23         24    356.6s
  24      3200   -0.547676709544736   -9.785e-03       24         24    714.1s
  24      6400   -0.545557804800340   +2.119e-03       24         24   1425.3s
```

- **χ=4 is the only χ that converges.** Bit-identical (`+0.000e+00`) across a 32× budget increase, so it reaches a fixed point in under 100 sweeps. It lands at −0.2719 — inside the bound, but far from the others.
- **Everything else is still moving at 6400 sweeps**: +2.1e-3 (24), +3.2e-3 (16), −5.8e-3 (6), and **+5.1e-2 (12)**, which is *worse* at 6400 than at 3200.
- Tightening `conv_tol` from 1e-10 to 1e-12 changes nothing, so these sweeps are terminating on `max_iter`, not on their convergence criterion.

## What this is not

**Not the #723/#726/#747 rank-1 collapse.** `ctm_corner_rank` is full at essentially every point (4, 6, 8, 12, 16, 24); the only exceptions are two mild deficiencies, 15/16 and 23/24. `env_is_collapsed` is false throughout, so the χ_eff=1 mean-field story does not apply. This is a *different fixed point / no fixed point*, not a degenerate one — closer in spirit to #833 (`ctm_tensor_c4v` landing differently from `ctm_tensor` on ~1/3 of D=3 states).

**Not state drift.** The simple-update state is bit-identical across all six χ (max elementwise difference exactly `0.0`, both sublattice tensors) — SU truncates at `max_bond_dim` and never reads χ, so χ enters only through the environment. Verified.

## Why it matters beyond this state

This is the root cause of #836. That issue's frozen energy literal had never once matched its own generator, and could not have: it was a sample of exactly this non-converged sweep. PR #837 removed the literal rather than refreshing it, because there is no stable number to freeze while this stands.

The state is `tests/_su_fixtures.physical_su_d2` — the fixture #772, #778, #784 and #785 all build on. Its *tensor* is sound and unaffected; what is affected is any energy computed from it through this path.

## Reproduce

```python
import jax; jax.config.update("jax_enable_x64", True)
from tenax.algorithms.ipeps import heisenberg_gate, ipeps, sublattice_rotate_gate
from tenax.algorithms.ipeps_config import CTMConfig, iPEPSConfig
from tenax.algorithms.ipeps_ctm_convergence import ctm_2site
from tenax.algorithms.ipeps_rdm import compute_energy_ctm_2site
from tenax.algorithms._ctm_diagnostics import ctm_corner_rank

gate = sublattice_rotate_gate(heisenberg_gate())
_E, (A, B), _ = ipeps(gate, None, iPEPSConfig(
    max_bond_dim=2, num_imaginary_steps=40, dt=0.05, unit_cell="1x1",
    ctm=CTMConfig(chi=6, max_iter=100, conv_tol=1e-10)))
A_d, B_d, g_d = A.todense(), B.todense(), gate.todense()

for chi in (4, 6, 12, 16):
    for mi in (400, 1600, 3200, 6400):
        eA, eB = ctm_2site(A_d, B_d, CTMConfig(chi=chi, max_iter=mi, conv_tol=1e-12))
        E = float(compute_energy_ctm_2site(A_d, B_d, eA, eB, g_d, 2))
        flag = "  <-- BELOW -0.6694" if E < -0.6694 else ""
        print(f"chi={chi:<3} max_iter={mi:<5} E={E:.15f} rank={ctm_corner_rank(eA)}{flag}")
```

## Caveats

- **One state**, not a scan over states. Whether this is specific to this rank-poor fixture or general to the 2-site CTM is untested.
- The "χ=4 gives about half" observation is approximate (ratio 0.502–0.511 depending on which non-converged value you divide by) and none of those divisors is a legitimate reference. I would not build on it.
- Aside, not a correctness claim: cost jumps ~160× between χ=16 and χ=24 at equal sweeps (8.9s → 1425.3s) where χ³D⁶ predicts ~3.4×. Likely memory-bound.

## Suggested first step

Determine whether the sweep is oscillating between two fixed points or wandering. Logging E per sweep for a few hundred iterations at χ=12 (the worst residual) would distinguish a limit cycle from drift, and costs seconds.

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Worth stating explicitly, since it narrows the search a lot: **this is forward-only. No AD is involved anywhere.**

The repro calls `ctm_2site` (plain forward convergence) and `compute_energy_ctm_2site`. No `grad`, no `value_and_grad`, no adjoint solve, no projector VJP. Dense tensors, D=2, trivial charges.

So none of the implicit-AD machinery is implicated: not the fixed-point backward, not GMRES/Neumann, not the rank clamp, not the gauge fixing. The forward sweep itself is failing to reach a fixed point.

That matters for everything built on top. Implicit AD *presumes* a converged forward — the adjoint solves `(I − Jᵀ)λ = dE/denv` at a fixed point that, on this state, does not exist at these budgets. Any gradient taken here is differentiating a point on an orbit, not a fixed point.

Two things about it are genuinely counterintuitive:

- **χ=4 converges and χ=6 does not.** More environment should make the fixed point easier to represent, not harder. The one χ that reaches a fixed point is the smallest one tried.
- **Non-monotone in budget.** χ=16 is inside the variational bound at 400 sweeps, outside it at 1600, back inside at 3200. More iterations makes it worse before better, which rules out "just slow to converge".

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

## Retracted — closing. The headline claims here are wrong.

Every dramatic result in the original report is an artifact of the **state**, not the CTM. The fixture is 40 simple-update steps at `dt=0.05`, i.e. imaginary time **τ=2.0**, which is not converged. I used the committed fixture config and never checked that.

Re-running the identical χ ladder across SU step counts:

```
=== tau=2 (as filed) ===                         max|dE|   settled  breach
   4   -0.271939907383136                        0.00e+00    YES       -
   6   -0.0549 .. -0.5747                        5.20e-01    no        -
   8   -0.5399 .. -0.5318                        3.44e-02    no        -
  12   -0.4050 .. -0.4965                        1.34e-01    no        -
  16   -0.5371 .. -1.0068                        4.70e-01    no       YES

=== tau=10 ===                        ||A-A_prev||_max = 3.61e-02
   4   -0.548635274239915                        0.00e+00    YES       -
   6   -0.5486                                   8.66e-06    no        -
   8   -0.5486                                   1.95e-05    no        -
  12   -0.5486                                   2.69e-05    no        -
  16   -0.5486                                   2.42e-05    no        -

=== tau=50 ===                        ||A-A_prev||_max = 7.01e-04
   4   -0.548637598963430                        0.00e+00    YES       -
   6..16  all -0.5486                  1.91e-05 .. 1.33e-04  no        -

=== tau=250 ===                       ||A-A_prev||_max = 9.16e-11
   4   -0.548637598963431                        0.00e+00    YES       -
   6..16  all -0.5486                  3.22e-05 .. 1.58e-04  no        -
```

By τ=10 every claim in this issue has evaporated:

- **The variational breach is gone.** No χ at any τ≥10 goes below −0.6694. The −1.0068 and −0.6967 points do not recur.
- **The factor of two is gone.** χ=4 gives −0.5486 at τ≥10, identical to every other χ. It was never a χ effect.
- **The order-of-magnitude oscillation is gone.** 5.2e-01 at τ=2 becomes 8.7e-06 at τ=10 — nearly five orders.

The SU state is converged by the end of the scan (‖A−A_prev‖ = 9.16e-11 between τ=50 and τ=250), so this is not a case of needing yet more imaginary time.

## What I got wrong, specifically

I treated `tests/_su_fixtures.physical_su_d2` as a representative physical state because it is the committed fixture. It is the opposite of representative **by design**: it exists for #772 precisely because its half-infinite environment is rank-poor and violates the retained-spectrum precondition. Building a general claim about the CTM on the one state selected for being pathological was the error, and "physical D=2 state" in the title was doing unearned work — it is a *physical* state, but an unconverged one.

## Two observations that survive, recorded here rather than left implied

Neither justifies keeping this issue open under its current claim. Filing separately if either matters:

**1. A residual ~1e-5 – 1.6e-4 floor for χ≥6 that does not improve with τ.** τ=10 is the *best* block (8.66e-06) and τ=250 is worse (1.58e-04) despite a state converged to 9e-11 — so it is not state quality, and `max|dE|` is already measured across budgets 400/1600/6400, so it is not sweeps either. χ=4 converges exactly (`0.00e+00`) at every τ.

There is a cheap hypothesis: D=2 makes the double-layer bond D²=4, so χ=4 is the only value with **no truncation**. That predicts χ=9 is the exactly-converging one at D=3 — one run to falsify, and worth doing before anyone theorizes further.

**2. `ctm_2site` discards its own convergence flag** (`ipeps_ctm_convergence.py`). The `while_loop` carry computes `converged` and the iteration count and both are dropped:

```python
converged = jnp.maximum(diff_A, diff_B) < conv_tol
...
env_A, env_B, _, _, _, _ = jax.lax.while_loop(cond_fn, body_fn, init_carry)
return env_A, env_B
```

`ipeps()` therefore cannot distinguish a converged environment from one that exhausted `max_iter`. That is the defect class of #801/#824. It is also why tightening `conv_tol` from 1e-10 to 1e-12 changed nothing in the original scan — the criterion never fired and the loop always ran to the cap, silently. Independent of everything above.

## #836 / PR #837 are unaffected

That fix stands, and for the same reason: the fixture's energy at χ=6 genuinely is not a freezable quantity, so removing the literal rather than regenerating it was correct. What changes is the scope of the explanation — it is a statement about that fixture, not about the 2-site CTM in general.

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Observation 2 from the retraction above is now filed separately as #839 — the dense CTM entry points discard their `converged` flag and iteration count, so `ipeps()` cannot tell a converged environment from an exhausted `max_iter`. It is independent of the retracted claim and survives it.

Observation 1 (the ~1e-5–1.6e-4 residual floor for χ≥6, and the D²=4 no-truncation hypothesis predicting χ=9 at D=3) remains unfiled.
